### ISSUE TITLE
Coverage report: Too many nested template calls

### ISSUE THREAD

**Issue by AirQuick** [xspec/xspec#215]
While analyze-string'ing an XSLT stylesheet, `coverage-report.xsl` makes so many recursive calls: https://github.com/xspec/xspec/blob/6e46603d5aa8406987dad00cab63bf5db841ab9e/src/reporter/coverage-report.xsl#L229

Thus the code coverage easily ends up with stack overflow as demonstrated by the attached test files:
[test.zip](https://github.com/xspec/xspec/files/2093723/test.zip)

```
C:\test>bin\xspec.bat -c test.xspec
...
Formatting Report...
passed: 1 / pending: 0 / failed: 0 / total: 1
Error at char 19 in xsl:variable/@select on line 310 column 46 of coverage-report.xsl:
  SXLM0001: Too many nested function calls. May be due to infinite recursion
  at xsl:apply-templates (.../src/reporter/coverage-report.xsl#278)
     processing /xsl:stylesheet/xsl:template[41]/text()[8]
  at test:coverage() (.../src/reporter/coverage-report.xsl#274)
  at xsl:call-template name="test:output-lines" (.../src/reporter/coverage-report.xsl#229)
  at xsl:call-template name="test:output-lines" (.../src/reporter/coverage-report.xsl#229)
  at xsl:call-template name="test:output-lines" (.../src/reporter/coverage-report.xsl#229)
...
```

**Comment by galtm** [comment]
Hi! I see that you have a proposed fix that requires XSLT 3.0. Before I was aware of that, I had obtained an XSLT 2.0 fix suggested by Michael Kay of Saxonica. It uses tail-recursion to avoid the stack overflow. Are you interested in the code for the XSLT 2.0 fix? If so, do you prefer that I post the code or do a pull request? (I'm new to GitHub, so I apologize for any lapses in process or etiquette.)

**Comment by galtm** [comment]
Sure. Here it is: https://github.com/galtm/xspec/pull/1

**Comment by AirQuick** [comment]
@galtm 
Thanks for the pull request!
(It is opened in [your forked repository](https://github.com/galtm/xspec/), though. We can't handle it unless opened here in [this repository](https://github.com/xspec/xspec/).)

The change looks promising. I tested it on Saxon-EE 9.7.0.21J with the test files attached to the initial message and indeed the problem didn't happen.

However, the same test on 9.8.0.12 failed:

```
$ export SAXON_CP=SaxonEE9-8-0-12J/saxon9ee.jar 
$ xspec-galtm-patch-1/bin/xspec.sh -c test/test.xspec 
Saxon script not found, invoking JVM directly instead.

Creating Test Stylesheet...

Running Tests...
Collecting test coverage data; suppressing progress report...
****************************************
controller=net.sf.saxon.Controller@55342f40

Formatting Report...
passed: 1 / pending: 0 / failed: 0 / total: 1
Error on line 124 column 53 of coverage-report.xsl:
  Too many nested template or function calls. The stylesheet may be looping.
  at xsl:apply-templates (file:/.../xspec-galtm-patch-1/src/reporter/coverage-report.xsl#85)
     processing /xsl:stylesheet
  at xsl:apply-templates (file:/.../xspec-galtm-patch-1/src/reporter/coverage-report.xsl#65)
     processing /
Too many nested template or function calls. The stylesheet may be looping.

*** Error formating the coverage report
```

I don't know if this is a known regression of Saxon 9.8 or not.
Anyway maybe this suggests that relying on a tail call optimization could be too implementation dependent. What do you think?

**Comment by galtm** [comment]
Good catch regarding Saxon 9.8. I was able to reproduce that error and sent details to Saxonica.

Yes, there is a certain fragility in relying on tail call optimization, and maybe it's not worth it. If your XSLT 3.0 solution does not require the stylesheet whose coverage is being analyzed to also use XSLT 3.0 (and my guess is that is does not), then that is probably the better approach.

(Thanks for the correction about where my pull request goes. I'll correct that within a few days.)

**Comment by galtm** [comment]
I think I fixed the pull request: https://github.com/xspec/xspec/pull/224
I haven't heard back from Saxonica yet, but I will keep you posted.

**Comment by galtm** [comment]
Michael Kay investigated the error in v9.8 and marked it Resolved. The bug record is marked "Applies to branch: 9.8, trunk" and "Fix Committed on: 9.8, trunk". Seems like the fix has not been included in a maintenance release yet, but it will be eventually.

**Comment by galtm** [comment]
One other thing -- Michael Kay provided a workaround for the version that produces the error: "disable bytecode generation using -opt:-c on the command line, (or various API options from Java)"

**Comment by AirQuick** [comment]
Thanks for the update, @galtm
I confirmed that #224 works fine with `-opt:-c` on Saxon-EE 9.8.0.14.
As soon as the next maintenance gets released, I'll see if #224 works on it without `-opt:-c`. If it works, I'll review the details and then we can probably merge it.

**Comment by AirQuick** [comment]
Ah, I think we can take the pull request without waiting for the next maintenance release. I'll start reviewing.

**Comment by AirQuick** [comment]
fixed by #224
